core/App/ehsm_napi.cpp

```diff
@@ -1155,15 +1160,60 @@ char *NAPI_RA_HANDSHAKE_MSG0(const char *p_msg0)
     log_d("***NAPI_RA_HANDSHAKE_MSG0 start.");
 
     log_d("msg0: \n %s", p_msg0);
-    std::string challenge_base64 = "challenge_base64";
-    std::string ga_base64 = "ga_base64";
-
     RetJsonObj retJsonObj;
+    ehsm_status_t ret = EH_OK;
+    sgx_ra_msg1_t *p_msg1;
+
+    memset(&p_msg1, 0, sizeof(p_msg1));
+
+    std::string challenge;
+    if (p_msg0 != nullptr)
+    {
+        Json::Value msg0_json;
+        Json::CharReaderBuilder builder;
+        Json::CharReader *reader(builder.newCharReader());
+        std::string err;
+        std::string response = p_msg0;
+        if (reader->parse(response.c_str(), response.c_str() + response.size(), &msg0_json, &err))
+        {
+            challenge = msg0_json["challenge"].asString();
+        }
+    }
+    else
+    {
+        retJsonObj.setCode(retJsonObj.CODE_BAD_REQUEST);
+        retJsonObj.setMessage("paramter invalid.");
+        goto out;
+    }
+    if(challenge.empty()){
+        retJsonObj.setCode(retJsonObj.CODE_BAD_REQUEST);
+        retJsonObj.setMessage("paramter invalid.");
+        goto out;
+    }
+
+    retJsonObj.addData_string("challenge", challenge);
+
+    p_msg1 = (sgx_ra_msg1_t *)malloc(sizeof(sgx_ra_msg1_t));
+    if (p_msg1 == NULL) {
+        retJsonObj.setCode(retJsonObj.CODE_FAILED);
+        retJsonObj.setMessage("Server exception.");
+        goto out;
+    }
+
+    ret = ra_get_msg1(p_msg1);
+    if (ret != EH_OK)
+    {
+        retJsonObj.setCode(retJsonObj.CODE_FAILED);
+        retJsonObj.setMessage("Server exception.");
+        goto out;
+    }
+
+    retJsonObj.addSubData_uint8Array("g_a", "gx", p_msg1->g_a.gx, SGX_ECP256_KEY_SIZE);
+    retJsonObj.addSubData_uint8Array("g_a", "gy", p_msg1->g_a.gy, SGX_ECP256_KEY_SIZE);
```

Comment:
still don't like this way to transfer the msg, it would be much more difficult if you want to refine the codes later.